### PR TITLE
Use reverse iteration protocol in `mapfoldr_impl`

### PR DESCRIPTION
**Pull request by YingboMa**
This PR addresses the issue #24823. I replace the `getindex` and `endof` by the reverse iteration protocol in the function `mapfoldr_impl`.

### CODE REVIEW

File: base/reduce.jl [hunk 1]
```diff
@@ -96,16 +96,16 @@ foldl(op, itr) = mapfoldl(identity, op, itr)
 
 ## foldr & mapfoldr
 
-function mapfoldr_impl(f, op, v0, itr, i::Integer)
+function mapfoldr_impl(f, op, v0, itr_r, i)
```
Comment (vchuravy) [inline base/reduce.jl]: Thanks for tackling this! I was wondering if there is any difference between `mapreduce`  and `mapfoldr_impl`?

```
mapfoldr(f, op, v0, itr) = mapreduce(f, op, v0 , Iterators.reverse(itr))
```

should be equivalent.

Comment (stevengj) [reply base/reduce.jl]: Two problems with this.

1. `mapreduce` doesn't define its associativity — it is allowed to associate however it wants.
2. even if you use `mapfoldl`, you are assuming that `op` is commutative as well as associative.  `foldr` changes associativity, but does not commute operands.

**Comment by stevengj** [comment]
Maybe add a test case for something that fails in 0.6, i.e. for an iterator that is reversible but not indexable.

For example,
```jl
@test foldr(Pair, 0, Iterators.filter(iseven, 1:6)) === foldr(Pair, 0, 2:2:6) === (2=>(4=>(6=>0)))
```

File: base/reduce.jl (hunk 1 above)
Comment (YingboMa) [reply base/reduce.jl]: There is a difference, if I am not mistaken.

```
julia> import Base.mapfoldr; mapfoldr(f, op, v0, itr) = mapfoldl(f, op, v0 , Iterators.reverse(itr))
mapfoldr (generic function with 2 methods)

julia> mapfoldr(identity, /, 1, 1:3)
0.16666666666666666
```
But the correct value should be 1.5.

**Review by stevengj — APPROVED** [review]

**Comment by KristofferC** [comment]
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/24833#issuecomment-348001470) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/c129101d71285ddf8310942c74d2841a84a97802/c45c820_vs_a73f9c9/report.md). cc @ararslan

**Comment by stevengj** [comment]
The performance regressions reported by Nanosoldier can't possibly be real; they are in lots of functions (e.g. `transpose!`) that have nothing to do with this PR.

**Comment by KristofferC** [comment]
This noise level is nothing out of the ordinary. Benchmarks looks good to me.

**Comment by stevengj** [comment]
±45% noise, with ±25% on dozens of benchmarks, is far worse than it used to be.  It makes nanosoldier almost useless.

**Comment by stevengj** [comment]
Why was this closed?

**Comment by StefanKarpinski** [comment]
Unclear. @YingboMa?

**Comment by YingboMa** [comment]
Oops sorry, I thought that this PR wasn't planned to be merged.

**Pull request closed without merging**